Reset CoreDataManager in the background and optimize logout

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14153 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR addresses some optimizations for logout:

- Deleted all Core Data managed objects in the background context. This ensures there is no discrepancy between the view context and the background context.
- Additionally, upon reset, `CoreDataManager` also triggers a notification for the reset, which causes the existing result controllers to fetch data again. We have places in the dashboard where we refresh data upon storage reset. To avoid refreshing data in the unauthenticated state, a few checks have also been added for orders, Blaze, and store onboarding cards.
- Added a change to `DefaultStoresManager` to reset payments only if the store is authenticated. This is to avoid an unexpected state where the user is unauthenticated but still has a default store. In #11373, the `AppCoordinator` would deauthenticate the app again in this state, triggering an unnecessary reset for `CardPresentPaymentAction` again.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->
- Log in to a test store.
- Navigate to Menu > Settings > Log out.
- Confirm logout and check in Xcode console to ensure that there is no more warning about "No processor found for ActionName".
- Confirm that the authenticator is displayed. 
- Log in again and confirm that all four tabs are loaded as expected.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->
Tested on simulator iPhone 16 Pro iOS 18.0 and confirmed that logout and login work as expected. The app functions normally after logging in again.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
N/A

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

TODO: @itsmeichigo to update the release notes before merging.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

itsmeichigo committed:
Handle reset CoreDataManager in the background

itsmeichigo committed:
Avoid reloading data on the dashboard when logging out

itsmeichigo committed:
Check authentication state before reloading dashboard cards

itsmeichigo committed:
Reset sessionManager before updating state to avoid leaving unauthenticated state with default store

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/229">20.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Revert "Check authentication state before reloading dashboard cards"

This reverts commit 2f9b95acf25e6f40c388ddc8f47c36e3bd793fa1.

itsmeichigo committed:
Fix failed unit tests

itsmeichigo committed:
Revert change on BlazeCampaignDashboardViewModel

itsmeichigo committed:
Avoid triggering data reload upon logout

itsmeichigo committed:
Add completion closure to reset method to fix unit test failure

itsmeichigo committed:
Merge branch 'trunk' into task/14153-reset-coredatamanager-in-background

itsmeichigo committed:
Fix UI test failure

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductsViewController.xib:
```diff
@@ -42,9 +42,6 @@
                         </view>
                         <tableView clipsSubviews="YES" contentMode="scaleToFill" alwaysBounceVertical="YES" style="plain" separatorStyle="default" rowHeight="-1" estimatedRowHeight="-1" sectionHeaderHeight="28" sectionFooterHeight="28" translatesAutoresizingMaskIntoConstraints="NO" id="1mE-SE-uK9">
                             <rect key="frame" x="0.0" y="51" width="414" height="797"/>
-                            <userDefinedRuntimeAttributes>
-                                <userDefinedRuntimeAttribute type="string" keyPath="accessibilityIdentifier" value="orders-table-view"/>
-                            </userDefinedRuntimeAttributes>
```
I'm not sure when this accessibility identifier has been wrongly set for the products list, but only just now that a UI test fails due to this. Removing the identifier fixes it.

itsmeichigo committed:
Merge branch 'trunk' into task/14153-reset-coredatamanager-in-background

itsmeichigo committed:
Restore deletion of objects on the view context

itsmeichigo committed:
Save view context after deleting objects

itsmeichigo committed:
Update unit test for reset

selanthiraiyan commented:
> Additionally, upon reset, CoreDataManager also triggers a notification for the reset, which causes the existing result controllers to fetch data again. We have places in the dashboard where we refresh data upon storage reset. To avoid refreshing data in the unauthenticated state, a few checks have also been added for orders, Blaze, and store onboarding cards.

Was this happening previously as well? Can we find a solution to avoid adding the authenticated check in different places? I am thinking that this is something that we can miss when we use the `onDidResetContent` in future. What do you think?

itsmeichigo commented:
> Was this happening previously as well?

This is already happening - you can see the log in Xcode in `trunk`.

> Can we find a solution to avoid adding the authenticated check in different places? I am thinking that this is something that we can miss when we use the onDidResetContent in future. What do you think?

This is my concern as well. In theory, it would be great if we don't trigger [refetching](https://github.com/woocommerce/woocommerce-ios/blob/2077debb97b5020c3aa1faaece30961beabf73e8/Yosemite/Yosemite/Tools/ResultsController.swift#L288) if the user is unauthenticated. However, `ResultController` doesn't have access to the stores manager so it's possible for now.

@selanthiraiyan  If you don't find this a blocker, we can create a subtask to make this better in the future. Otherwise, I can revert the changes and try to come up with a better solution later.

selanthiraiyan commented:
> @selanthiraiyan If you don't find this a blocker, we can create a subtask to make this better in the future. Otherwise, I can revert the changes and try to come up with a better solution later.

Thanks for the explanation! I am good with creating a subtask. Thank you!

selanthiraiyan reviewed (approved):
This tests as described, and the code changes LGTM. 

I agree that we can deal with "the authenticate check" in a sub task. 👍

itsmeichigo committed:
Merge branch 'trunk' into task/14153-reset-coredatamanager-in-background

itsmeichigo committed:
Update release notes

itsmeichigo commented:
Thanks! I added a checkbox in #14050 for the suggested improvement. Also updated the release notes in 5cb8a282650c6389f294606daaf987c1ee00faec. Enabling auto merge now.
